File: python/open_guardrail/guards/internal_url_detect.py
"""Detect internal/private URLs in output."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"https?://localhost(:\d+)?", re.IGNORECASE
    ),
    re.compile(
        r"https?://127\.0\.0\.1(:\d+)?",
        re.IGNORECASE,
    ),
    re.compile(
        r"https?://0\.0\.0\.0(:\d+)?", re.IGNORECASE
    ),
    re.compile(
        r"https?://10\.\d{1,3}\.\d{1,3}\.\d{1,3}"
    ),
    re.compile(
        r"https?://172\.(1[6-9]|2\d|3[01])"
        r"\.\d{1,3}\.\d{1,3}"
    ),
    re.compile(
        r"https?://192\.168\.\d{1,3}\.\d{1,3}"
    ),
    re.compile(
        r"https?://[a-zA-Z0-9-]+\.local\b"
    ),
    re.compile(
        r"https?://[a-zA-Z0-9-]+\.internal\b"
    ),
    re.compile(
        r"https?://[a-zA-Z0-9-]+\.corp\b"
    ),
    re.compile(r"https?://\[::1\]"),
]


class _InternalUrlDetect:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "internal-url-detect"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="internal-url-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Internal URL detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def internal_url_detect(
    *, action: str = "block"
) -> _InternalUrlDetect:
    return _InternalUrlDetect(action=action)

File: packages/guards/src/internal-url-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface InternalUrlDetectOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /https?:\/\/localhost(:\d+)?/i,
  /https?:\/\/127\.0\.0\.1(:\d+)?/i,
  /https?:\/\/0\.0\.0\.0(:\d+)?/i,
  /https?:\/\/10\.\d{1,3}\.\d{1,3}\.\d{1,3}/,
  /https?:\/\/172\.(1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3}/,
  /https?:\/\/192\.168\.\d{1,3}\.\d{1,3}/,
  /https?:\/\/[a-zA-Z0-9-]+\.local\b/,
  /https?:\/\/[a-zA-Z0-9-]+\.internal\b/,
  /https?:\/\/[a-zA-Z0-9-]+\.corp\b/,
  /https?:\/\/\[::1\]/,
];

export function internalUrlDetect(options: InternalUrlDetectOptions): Guard {
  return {
    name: 'internal-url-detect',
    version: '0.1.0',
    description: 'Detects internal/private URLs in output',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'internal-url-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
